packages/backend/src/tools/changelog/diff.ts

```diff
@@ -0,0 +1,170 @@
+import { assert } from '@l2beat/backend-tools'
+import { ContractParameters, DiscoveryOutput } from '@l2beat/discovery-types'
+import { ChainId, EthereumAddress } from '@lz/libs'
+import diff from 'deep-diff'
+
+import { getMatchingContractPairs, splitSafeContractPairs } from './pairs'
+import { ChangelogEntry, FieldDifference } from './types'
+
+export {
+  changelogEntryToFieldDiff,
+  diffContractValues,
+  fieldDiffToChangelogEntry,
+  getDiscoveryChanges,
+}
+
+function getDiscoveryChanges(
+  previousOutput: DiscoveryOutput,
+  currentOutput: DiscoveryOutput,
+): ChangelogEntry[] {
+  const matchingPairs = getMatchingContractPairs(previousOutput, currentOutput)
+
+  const { safe: safePairs } = splitSafeContractPairs(matchingPairs)
```

Comment:
So we should do something with unsafe pairs. We should store information that contract appeared/disappeared from the system.

Reply:
Will be a part of changelog indexer - already implemented on draft PR